internal/controller/apiextensions/definition/reconciler.go

```diff
@@ -485,147 +482,124 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
+			watch.WithLogger(log),
+			watch.AlwaysKeep(xrGVK, v1.CompositionRevisionGroupVersionKind),
+		)
+		co = append(co, engine.WithWatchGarbageCollector(gc))
 	}
 
-	c, err := r.composite.Create(name, ko, watches...)
-	if err != nil {
+	if err := r.controller.Start(name, co...); err != nil {
 		log.Debug(errStartController, "error", err)
 		err = errors.Wrap(err, errStartController)
 		r.record.Event(d, event.Warning(reasonEstablishXR, err))
 		return reconcile.Result{}, err
 	}
 
-	if r.options.Features.Enabled(features.EnableAlphaRealtimeCompositions) {
-		ca = c.GetCache()
-		if err := ca.IndexField(ctx, u, compositeResourceRefGVKsIndex, IndexCompositeResourceRefGVKs); err != nil {
-			log.Debug(errAddIndex, "error", err)
-			// Nothing we can do. At worst, we won't have realtime updates.
-		}
-		if err := ca.IndexField(ctx, u, compositeResourcesRefsIndex, IndexCompositeResourcesRefs); err != nil {
-			log.Debug(errAddIndex, "error", err)
-			// Nothing we can do. At worst, we won't have realtime updates.
-		}
-	}
-
-	if err := c.Start(context.Background()); err != nil { //nolint:contextcheck // the controller actually runs in the background.
-		log.Debug(errStartController, "error", err)
-		err = errors.Wrap(err, errStartController)
+	if err := r.controller.StartWatches(name,
+		engine.WatchFor(u, &handler.EnqueueRequestForObject{}),
+		engine.WatchFor(&v1.CompositionRevision{}, EnqueueForCompositionRevision(resource.CompositeKind(xrGVK), r.client, log)),
+	); err != nil {
+		log.Debug(errStartWatches, "error", err)
+		err = errors.Wrap(err, errStartWatches)
 		r.record.Event(d, event.Warning(reasonEstablishXR, err))
 		return reconcile.Result{}, err
 	}
-	log.Debug("(Re)started composite resource controller")
 
-	if r.options.Features.Enabled(features.EnableAlphaRealtimeCompositions) {
-		r.xrInformers.RegisterComposite(xrGVK, ca)
-	}
+	log.Debug("Started composite resource controller")
 
 	d.Status.Controllers.CompositeResourceTypeRef = v1.TypeReferenceTo(d.GetCompositeGroupVersionKind())
 	d.Status.SetConditions(v1.WatchingComposite())
 	return reconcile.Result{Requeue: false}, errors.Wrap(r.client.Status().Update(ctx, d), errUpdateStatus)
 }
 
-func (r *Reconciler) stopCompositeController(d *v1.CompositeResourceDefinition) {
-	r.composite.Stop(composite.ControllerName(d.GetName()))
-	if r.options.Features.Enabled(features.EnableAlphaRealtimeCompositions) {
-		r.xrInformers.UnregisterComposite(d.GetCompositeGroupVersionKind())
-	}
-}
-
 // CompositeReconcilerOptions builds the options for a composite resource
 // reconciler. The options vary based on the supplied feature flags.
-func CompositeReconcilerOptions(co apiextensionscontroller.Options, d *v1.CompositeResourceDefinition, c client.Client, l logging.Logger, e event.Recorder) []composite.ReconcilerOption {
+func (r *Reconciler) CompositeReconcilerOptions(ctx context.Context, d *v1.CompositeResourceDefinition) []composite.ReconcilerOption {
```

Comment:
I made this a method because I noticed:

* `Reconciler` was the only caller of the function
* Almost all of the arguments were fields on `Reconciler`
* I wanted to add more arguments that were also fields on `Reconciler`, and I wasn't happy with how long the function signature was getting

Reply:
nit: worth its own commit.